src/shammath/include/shammath/matrix_exponential.hpp

```diff
@@ -0,0 +1,423 @@
+                ld_9(s_tilde);
+            }
+        };
+
+        process_val(val);
+    }
+
+    /**
+     * @brief This function compute the Taylor polynomial up to order m_star
+     * @param q Paterson-Stockmeyer interger (it's used to compute the matrix power)
+     * @param r Paterson-Stockmeyer polynomial degree
+     * @param bi_seq sequence of coef needed for Paterson-Stockmeyer coefficient B_k
+     * @param size size of matrices
+     * @param A input matrix
+     * @param F output matrix
+     * @param B,I,Id matrices for intermediate computations
+     */
+    template<
+        typename T,
+        class U,
+        class Extents1,
+        class Extents2,
+        class Extents3,
+        class Extents4,
+        class Extents5,
+        class Layout1,
+        class Layout2,
+        class Layout3,
+        class Layout4,
+        class Layout5,
+        class Accessor1,
+        class Accessor2,
+        class Accessor3,
+        class Accessor4,
+        class Accessor5>
+    inline void taylor_eval(
+        const i32 r,
+        const i32 q,
+        std::array<f64, 30> &bi_seq,
+        const size_t size,
+        const std::mdspan<T, Extents1, Layout1, Accessor1> &A,
+        const std::mdspan<T, Extents2, Layout2, Accessor2> &F,
+        const std::mdspan<T, Extents3, Layout3, Accessor3> &B,
+        const std::mdspan<T, Extents4, Layout4, Accessor4> &I,
+        const std::mdspan<T, Extents5, Layout5, Accessor5> &Id) {
+        mat_set_nul<T>(F);
+
+        for (auto k = r - 1; k >= 0; k--) {
+            mat_set_identity<T>(I);
+            mat_set_identity<T>(Id);
+            mat_set_nul<T>(B);
+            i32 cc = 0;
+
+            for (auto j = 1; j <= q; j++) {
+                mat_copy<T>(I, Id);
+                mat_gemm<T, U>(1, A, Id, 0, I);
+                cc = q * k + j;
+                mat_axpy_beta<T, U>(bi_seq[cc], I, 1, B);
+            }
+            mat_set_identity<T>(Id);
+
+            i32 cond = (k >= 1);
+            mat_axpy_beta<T, U>(1, B, 1, F);
+            mat_axpy_beta<T, U>(1 - cond, Id, cond, I);
+
+            mat_gemm<T, U>(1, F, I, 0, B);
+            mat_copy<T>(B, F);
+        }
+        mat_plus_equal_scalar_id<T, U>(F, 1);
+    }
+
+    /**
+     * @brief matrix scaling-squaring Talylor-based matrix exponential
+     * @param K maximum number of matrix product allow
+     * @param A input matrix
+     * @param F output matrix
+     * @param B,I,Id matrices
+     * @param size_A size of matrices
+     */
+    template<
+        typename T,
+        class U,
+        class Extents1,
+        class Extents2,
+        class Extents3,
+        class Extents4,
+        class Extents5,
+        class Layout1,
+        class Layout2,
+        class Layout3,
+        class Layout4,
+        class Layout5,
+        class Accessor1,
+        class Accessor2,
+        class Accessor3,
+        class Accessor4,
+        class Accessor5>
+    inline void mat_expo(
```

Comment:
```suggestion
    inline void mat_exp(
```
In general the abreviation is always exp instead of something longer for exponentials